### ISSUE TITLE
[HOLD for payment 2023-03-31] Implement New Date Picker Design

### ISSUE THREAD

**Issue by Beamanator** [Expensify/App#14322]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
We decided internally back [here](https://expensify.slack.com/archives/C03U7DCU4/p1660248419773359) that we want to move forward with this kind of mindset for Dates:

> - **Single full date**: calendar date picker
> - **Date range**: independent Start and End date inputs, which pushes to the calendar date picker
> - **Partial date**: two text boxes with MM and YYYY placeholders. Since these two fields make up one data set, they would be shown together when pushed to an additional page.

Until now this has not been implemented, and now in this issue (https://github.com/Expensify/App/issues/11580) we created a new page for a user's Date of Birth, where there will only be 1 date field (Single full date). This is the a very empty page and would be a great time to start showing a full date picker

For now we'll only build the **Single full date: calendar date picker** - the rest can come later

Here's our "official" design we want to build: https://github.com/Expensify/App/issues/14322#issuecomment-1398381437

Bugs in existing date picker that should be solved in new date picker:
- https://github.com/Expensify/App/issues/14147
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01af33f93559a520a9</li>
        <li>Upwork Job ID: 1642945974459355136</li>
        <li>Last Price Increase: 2023-04-03</li>
    </ul>
</details>

**Comment by shawnborton** [comment]
So I think we don't really have any new learnings other than where we left off @Beamanator but one thing with this particular screen you are asking for (birthdate), I feel like having a date picker for a birth date is pointless. Do you really need a calendar to input your date of birth? Wouldn't it be more clean if we just had three separate input fields for it? 

Maybe that one is worth taking back to the room for discussion?

**Comment by zanyrenney** [comment]
@shawnborton I believe we landed on the date picker for situations where we're displaying the single full date (of which DOB would be one) [here](https://expensify.slack.com/archives/C03U7DCU4/p1660248419773359) and then Jason mentioned again that Account Settings would be a good time to bring this into the wild [here](https://expensify.slack.com/archives/C02HWMSMZEC/p1660597941679799) so it feels like we did land on that option. 

Though, it has been a while (6 months 😅 ) so if you aren't convinced and think a discussion in n07 would be best, maybe that's a sensible approach @Beamanator - what do you think?

**Comment by shawnborton** [comment]
Yeah, for a birth day at least, I think it makes sense to bring this to the room again - I'm happy to lead that. 

The point being that if someone asked you what your birthday was, would you rather just tell them, or go get out 35+ years worth of calendars and then find the exact date in the calendar and show them?

**Comment by zanyrenney** [comment]
Ha, yeah fair challenge in the context of a conversation, in the context of a financial app and your DOB being asked though, a more complete approach might be more normal. Regardless - even this back and forth shows there are different reads of this 😀 so 👍🏼 to take this to the room. Thanks for offering to lead that conversation!

**Comment by shawnborton** [comment]
Actually maybe the picker will be fine... started a [conversation here](https://expensify.slack.com/archives/C01GTK53T8Q/p1673894240895279), let me know what you think.

**Comment by Beamanator** [comment]
Thanks for starting the convo in slack, @shawnborton ! 👍

**Comment by kevinksullivan** [comment]
Sorry catching up, but we went down the rabbit hole on different date inputs in [this thread](https://expensify.slack.com/archives/C03U7DCU4/p1660154351016469?thread_ts=1659044343.913889&cid=C03U7DCU4), and concluded we'd go with a date picker for all full dates to start. The frontrunner from that thread was this mock:

![image](https://user-images.githubusercontent.com/22508290/213588577-fd258a6d-8e05-4eb5-a746-202fddc68d25.png)



That's largely in line with what you posted the other day @shawnborton , aside from the down caret v. right caret pushing to years.

**Comment by kevinksullivan** [comment]
@Beamanator do we only need to get clear on the full date pattern right now, or do we need to clarify date range and partial date as well?

**Comment by Beamanator** [comment]
@kevinksullivan thanks for posting that and I think we should try to get clear on everything everything before implementing here - Technically the Account Settings implementation isn't blocked on this, but N7 isn't complete till we finish this up so let's completely describe how it should work before we move to implementation 👍

**Comment by shawnborton** [comment]
I still feel good about this being our v1 datepicker:
![image](https://user-images.githubusercontent.com/2319350/213705301-556dcd3a-f5d6-4a1e-82c4-f0856230dc45.png)

**Comment by Beamanator** [comment]
I like this 👍  And just to be clear @shawnborton - do we want an X in the top right (vertically aligned with the "back" button) that can close out the entire modal?

**Comment by shawnborton** [comment]
Ah yes, that would make sense. I know we'll eventually get rid of that top right x (cc @puneetlath) but I guess I was getting ahead of ourselves!

**Comment by Beamanator** [comment]
Aah ok cool just making sure 🙏 👍

**Comment by Beamanator** [comment]
Ok I think the design is complete, I'm planning to make this `Weekly` since it's not highest highest priority

**Comment by luacmartins** [comment]
Should we add datepicker related bugs, e.g. [this one](https://github.com/Expensify/App/issues/14147) to this issue to make sure the new design solves any existing problems?

**Comment by Beamanator** [comment]
@luacmartins good idea 👍 I made a new section in the OP to start linking existing bugs that the new date picker needs to not have

And @JmillsExpensify yep good idea 👍

**Comment by Beamanator** [comment]
Added to N7 Project & removed N7 from title, will assign this out very soon

**Comment by ArekChr** [comment]
Hello, I can work on the proposal, callstack

**Comment by ArekChr** [comment]
### Problem
We want to implement a consistent date picker across all platforms (iOS, Android, web) for a new page for the user's Date of Birth, where there will only be one entire date field.

### Root Cause
The current solution needs more consistency across all platforms, which creates a difficult user experience.

### Solution
To create a consistent date picker, I suggest you implement a custom calendar. The calendar will be embedded in a modal that will display the calendar on all platforms. The user can switch between months by clicking on the left and right arrows. I can implement animations between transitions to make the calendar even more user-friendly.
To generate the calendar, I will use Moment.js to generate matrices of months for individual days and determine the starting day of the week. That will ensure the calendar is optimized and only renders the current month plus the previous and next month to add animations when switching months.
Alternative Solutions
None were explored.

### Implementation Steps
1. Use Moment.js to generate months matrices to determine the week's starting day.
2. Create a calendar component that displays a specific month for the current value or day if a value is not provided.
3. (optional) Create a lazy loaded previous and next calendar month (this is needed if we want to implement animation between switching months).
4. (optional) minimum and maximum date properties that prevent users from clicking dates outside the range.
5. (optional) the first day of the week determines the day considered the beginning of the week.
6. Implement the ability to switch between months by clicking on the left and right arrows.
7. Implement a simple year and month picker using FlatList.
8. Reuse the existing modal to display the calendar; on the web, I will use styles that show the calendar relative to the date input. On mobile, it will be centred by modal.
9. Test the implementation on all platforms (iOS, Android, web) to ensure consistency across all platforms.

### Conclusion
Implementing a custom calendar can ensure a consistent date picker experience across all platforms. The use of Moment.js will make generating the calendar more accessible and more performant. This solution will provide a better user experience for the new Date of Birth page.
This solution will also make it easier to add more date picker formats, such as full date and time.

**Comment by Beamanator** [comment]
@ArekChr overall I love the implementation plan 👍 

I do want to ask @shawnborton if we need the animation when switching months? Also @ArekChr what kind of animation are you proposing to display at that point?

**Comment by shawnborton** [comment]
Animation isn't necessarily needed, but it could be nice. cc @stitesExpensify as he has some experience using animations with our emoji picker projects. 

I could see where animating the months from left to right could look nice.

**Comment by ArekChr** [comment]
@Beamanator I can implement a simple swipe animation that will perfectly fit this component, so it will emit that the calendar is scrollable. It will give us low memory usage as we generate only three calendars, current: `n`, next: `n+1` and last: `n-1`. Changing the current month to the previous one will always unmount all calendars that are out of range for n > 1 and n < -1. Such functionality should give a good performance for low-end devices and provide good UX.

**Comment by Beamanator** [comment]
@ArekChr that's pretty interesting what you're talking about, but I'm wondering - is it much better to generate three calendars (current, next, and previous) instead of just having one calendar that regenerates / repopulates every time you click the next or previous month?

**Comment by ArekChr** [comment]
@Beamanator I assumed that it is good to have one previous and next generated calendar to prevent a delay between mounting the component and then starting the animation. I can do that in the way you suggested and test performance while implementing it to check if that solution is good enough

**Comment by Beamanator** [comment]
@ArekChr thanks for the response, sorry for my delay!

I think you're right that with swipe gestures we might need a pre-rendered calendar ready to slide into view while the existing one is sliding out of view, so I would say I like your plan of attack 👍

**Comment by ArekChr** [comment]
@Beamanator I started working on that issue together with @TMisiukiewicz

**Comment by TMisiukiewicz** [comment]
progress update: together with @ArekChr we were able to create the datepicker for iOS and Android:
<img width="348" alt="image" src="https://user-images.githubusercontent.com/13985840/218497724-03d909be-14c7-4483-a866-79ca16cfe86a.png">
The next steps are creating the same component for Web and month & yeah picker

**Comment by shawnborton** [comment]
Looking good so far! Mind tagging me in your PR and I can help provide design feedback when you are ready for it? Thanks!

**Comment by ArekChr** [comment]
#15119 linking an issue that will be fixed in the new date picker design

**Comment by TMisiukiewicz** [comment]
> Looking good so far! Mind tagging me in your PR and I can help provide design feedback when you are ready for it? Thanks!

No problem!

Today we managed to progress with year & month pickers and displaying the new picker component across all the platforms, here are some videos presenting our progress:

https://user-images.githubusercontent.com/13985840/218778767-b7a1ae89-cdcd-4c27-8488-8c2a31ab357c.mov


https://user-images.githubusercontent.com/13985840/218778774-e02128f1-2bb3-434a-9dad-4ef5b857d225.mov


https://user-images.githubusercontent.com/13985840/218778776-b5b9d3c3-8ebf-4ee9-9000-db3c52be008d.mov


https://user-images.githubusercontent.com/13985840/218778780-cdbfdbe3-4c5d-4b12-a5e8-ffbf12307a8b.mov


Providing the answers for the following questions will definitely help us with moving forward:

1. What should the default input date format be? (now is YYYY-MM-DD )
2. What should the border for month and year pickers look like? (can it be the same as it is on the calendar? It might look weird if we have a different background under the picker - you can see on the videos currently it has the same style)
3. What height should be for Month and Year pickers? right now is set to 510px (10 elements of the list are visible)
4. Do we want a support property that sets the start of the week? (right now starting day of the week is Sunday)

**Comment by JmillsExpensify** [comment]
Quickly jumping in on this one, though leaving the rest to @shawnborton.

> What should the default input date format be? (now is YYYY-MM-DD )

Yes we use YYYY-MM-DD throughout the app. We haven't internationalized dates yet.

**Comment by shawnborton** [comment]
For your first video, we should be pushing to a new page for those subsequent screens right? @JmillsExpensify @Beamanator was that your understanding as well?

**Comment by shawnborton** [comment]
So if we push to new pages for year and month, we don't need to worry about having "year picker" and "month picker" styles. They just use our standard push-to-page list selection style.

**Comment by Beamanator** [comment]
> For your first video, we should be pushing to a new page for those subsequent screens right? 

Yeah that was my understanding as well 👍  And I agree we won't really have to worry about the "Year picker" / "Month picker" styles in that case

@ArekChr @TMisiukiewicz you can see examples of this "push-to-page list selection style" by visiting these pages:
1. Settings -> Profile -> Pronouns
2. Settings -> Profile -> Timezone -> Timezone (this one has a textinput at the top which allows for searching)

**Comment by Beamanator** [comment]
> 4. Do we want a support property that sets the start of the week? (right now starting day of the week is Sunday)

I don't think we need that for v1, it could be useful in the future though - maybe later we can run a poll in #expensify-open-source to see if people really want that feature

**Comment by JmillsExpensify** [comment]
> For your first video, we should be pushing to a new page for those subsequent screens right? @JmillsExpensify @Beamanator was that your understanding as well?

Dang that's a really good catch. Agree with @Beamanator @shawnborton that we want to make sure this picker is consistent with the "push to page" pattern we already have in Settings > Profile.

**Comment by shawnborton** [comment]
Your understanding sounds correct for me. The text input in the list view when you push should work just as a search input to search the list.

**Comment by mountiny** [comment]
That is some great progress here @TMisiukiewicz and @ArekChr keep it up! Feel free to share a draft PR as well if you have some ready 🙌

**Comment by JmillsExpensify** [comment]
>
hey, some quick update from our side today:

we added minDate and maxDate props for web and desktop to prevent users selecting the out-of-range dates (on the screenshot there is a maximum date set to 2023-04-17):
image

- we refactored our code to match it with Expensify guides - we added tests to a function responsible for generating month calendar
So, in case of the picker, please tell us if we are correct:

when user has the DatePicker open, and presses the Month, we are using "push to page" pattern and move to the screen with Months list. After selecting the month or pressing "back", we are back on the screen with DatePicker component with calendar open and month selected
when user has the DatePicker open, and presses the Year, we are using "push to page" pattern and move to the screen with Years list. After selecting the year or pressing "back", we are back on the screen with DatePicker component with calendar open and year selected
What about those TextInputs above the lists on the design? Are they used to put a year/month manually or search through the list? I'm not sure if it makes sense to enter it manually since i don't see any confirmation button on the design and it might lead to some edge cases that we would need to cover, on the other hand these lists are pretty straighforward and don't have many items, so using this input to search through the records might not be needed as well

Manual input in the top-line date field shouldn't be allowed. Otherwise we have to handle all kinds of date formats. As for the push to page lists, @shawnborton are you saying that Year and Month will take search as a input? So typing `19` would filter out any 2000+ results?

**Comment by JmillsExpensify** [comment]
Ah sorry my bad, just didn't think of that. So to clarify though, that's what you are suggesting, right? We'll have search on month/year versus just scroll to start?

**Comment by shawnborton** [comment]
Well my thinking is that we'd just follow how this pattern works elsewhere right? Where if we have more than XX number of options in the list, we have a search box.

**Comment by JmillsExpensify** [comment]
Ahhh, ok yeah that's a good reminder, and a reason we should write this stuff down and put it in the /App repo – basically exactly like we've done for forms.

**Comment by ArekChr** [comment]
@shawnborton Referring to the method of selecting the year and month by push-to-page, I am concerned about the complexity of the code. If I separate the logic of the calendar component, I may encounter several drawbacks that could make your code harder to maintain and extend in the long term.

If I separate the calendar component into 2 screens (month list screen and year list screen), by push-to-page and the calendar itself, it could be harder to communicate with other parts of your application. For example, if I need to pass the selected month or year values to another screen, I must implement complex communication channels between the calendar and the other screens. This could lead to more complex code, more bugs, and slower development times.

Separating the calendar component from month/year pickers could be harder to reuse in other parts of your application. You could leverage existing functionality and maintain a more consistent codebase by keeping the component integrated with the rest of the application by making it hermetic. This could improve code reusability and reduce the time and effort required to build new features or applications.

**Comment by mountiny** [comment]
@ArekChr thanks for articulating the concerns, I think they are valid, but in this case, the date picker is not a component which would  be used in many places of the App. AFAIK there are only handful (3) places where we would use it hence the reusability is not super worrying to us I would say.

What drives this design is a desire for uniform design and patterns in the app so this is easy to understand for any user. I think we should try to implement this, even keeping in mind the reusability is not that easy or straight forward. What do you think @Beamanator?

**Comment by JmillsExpensify** [comment]
Agreed, thanks for raising those concerns. Let's see what @Beamanator thinks when he returns, though the complexity of the picker does worry me, if I'm being honest. These is going to be used in one of the core flows in the app, which is adding a business bank account.

**Comment by ArekChr** [comment]
Hello, we're currently investigating the push-to-page implementation for the year and month picker and are working on the PR checklist. Our initial implementation is almost complete, and we have handled all edge cases and unit tests except for the year and month picker. We are waiting for your decision about whether to push the year and month screen to the page.
The DateInput, together with the year and month picker, is a straightforward and fully reusable component. It can be used in many places in the app without code changes, which is hugely advantageous. However, if you decide to change the implementation to push-to-page, please keep in mind that it may require more complex communication channel changes and take more time to refactor, as we would have to connect with the global store to set the date of birth in both the YearListScreen and MonthListScreen, which breaks encapsulation for the DateInput. It will be for one use case: the date of birth input.

**Comment by mountiny** [comment]
I am personally fine with not requiring the push-to-page here, I think the benefits of having this as one component are better and in terms of consistency its also arguable as this component is very unique and hence the main consistency we should care about is consistency of the datepicker across the platforms and not necessarily try to match behaviour to other elements.

I think UX wont be influenced in a bad way if we will not use push-to-page. @shawnborton @Beamanator @JmillsExpensify @luacmartins What are your thoughts on this?

**Comment by JmillsExpensify** [comment]
Interesting. I'm pretty torn, because I think when we started this journey, if we knew what we know now, it's possible that we wouldn't have gone with the design we have. Accordingly, I think I agree with @mountiny that we should get this improvement out in the wild, essentially merging what's already been coded. The main counterargument could be that breaking the encapsulation of the DateInput is fine because we'll also eventually use it for date of birth and perhaps incorporation date – but even then we're adding complication for two flows that you set once and never change. I'm finding it hard to justify the "consistency above all else" path as a result.

**Comment by shawnborton** [comment]
> However, if you decide to change the implementation to push-to-page

Well, we originally intended for push-to-page right? So we aren't changing to push-to-page, that was always what we had planned to have. 

> please keep in mind that it may require more complex communication channel changes and take more time to refactor

Can you elaborate on what exactly this means?

**Comment by ArekChr** [comment]
@shawnborton Yes, the CalendarPicker component currently updates its state when the user selects a new year or month from the ListPicker. It then calls the onChange function, and passes a new date through the onChange function in DateInput, which can be used outside the component, making it reusable. However, suppose we want to implement a "Push to page" functionality. In that case, we can't pass the onChange function between the CalendarPicker and the year/month Screen because we can't pass functions that are not serializable. Instead, we need to manage the state globally by pressing month or year and breaking communication between the calendar and month/year picker. We set the date of birth directly global state, breaking month and year picker reusability which is more like 'hacky' solution

**Comment by shawnborton** [comment]
Got it, I am not very technical so I will let our engineers respond to that. But I would be curious how we are currently handling push-to-page for things like our Settings page, etc? That seems to work just fine, but I might be missing something.

**Comment by JmillsExpensify** [comment]
I think those are different given that we don't have a picker in play yet. It's just a series of forms, right?

**Comment by Beamanator** [comment]
@ArekChr I was just thinking about this situation and I think in the future we're going to have similar situations to what we're dealing with in this picker. Basically my understanding is that we might, in the near future, break large forms into multiple push-to-page segments (correct me if I'm wrong @shawnborton / @JmillsExpensify ) so I do think we need some kind of solution how this will align with our push-to-page plan.

One idea I had for this is something like:
1. We already know the form onyx key that the date is a part of ([example](https://github.com/Expensify/App/blob/660e3892f8a8deace5e24497148a174ee8dde251/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js#L86-L88))
2. We could store that form onyx key & the field `inputID` in a new temporary Onyx key (like `ONYXKEYS.DATE_INFO_BEING_MODIFIED` (i'm pretty trash at coming up with onyx key names))
    1. We could do something like this right before redirecting to the year / month selection pages:
    ```js
    Onyx.merge(ONYXKEYS.DATE_INFO_BEING_MODIFIED, {
        formID, // the onyx key of the form being modified
        inputID, // the id of the date input in the form that's being modified
    });
    ```
3. When we push to a new page like year / month selector, we can read from that key just so we know:
    1. The default year / month to highlight
    2. What onyx key to update when a year / month is selected
4. When the year / month is selected (or when that page is closed), we:
    1. Update the form's date input with the new month / year that was selected
    2. Clear out `ONYXKEYS.DATE_INFO_BEING_MODIFIED`

How does that sound? Is that way over complicating things?

**Comment by TMisiukiewicz** [comment]
In my opinion it totally makes sense to break large forms into multiple push-to-page segments - filling a form is a process, and push-to-page is definitely helpful in this situation. However, picking a date is already part of this process - user can select the day from the calendar without even going to the Month or Year picker. Both of these pickers are related only to one field in a form. It means, to follow the push-to-page pattern, each datepicker in a form should have two additional pages and routes. Let’s dive into the easiest possible example:

Let’s say you have a page where you have personal information combined into one form. It’s under `settings/profile` route. Two of the inputs are date pickers - Date of birth and ID Expiry date. In this case, you have to create 4 new routes when you want to follow push-to-page pattern:
`settings/profile/dob/month`
`settings/profile/dob/year`
`settings/profile/idexp/month`
`settings/profile/idexp/year`

You need this, as using push-to-page pattern you want to know where exactly to go back when you are on the page, and which date picker you should keep opened after going back. You will be also able to enter this route directly, without even going into the form - just pasting url into the browser - which does not seem to make sense in case of datepicker. If datepicker is using min/max date, we have to pass it to the screen to block months and years out of the range. And it might be very dynamic - one year all months might be available, and another just few of them. As you can see it generates multiple of edge cases that need to be covered. And each time date picker is used, there is a need of creating new routes bound to it to make it work. With our proposed approach, all of this problems does not exist. I’m totally aware that push-to-page pattern seems to be a good idea with the date inputs currently existing in the app, but i think in this case we should bear in mind the bigger picture - every new date input in the app means implementing it depending on the navigation. Without navigation the picker is just a half-baked component, not possible to reuse.

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by shawnborton** [comment]
I'm still feeling good about doing push-to-page for all of this. I tried exploring some flows that are in line with what you all are building right now where the date picker just morphs into the month selector and the year selector, but it really gets weird when you try to select a year. 

For example, push-to-page is great for year selection because it gives us the search input above the giant list. If we try to recreate that with the inline replacement flow you made, we get a fairly awkward experience:
![image](https://user-images.githubusercontent.com/2319350/220361659-3419dc40-02a1-43d5-8e15-aa44b45975a2.png)


Also notice how I added headings with an "x" button to each of those replacement flows, as I think that would be a UX requirement so that a user could back out of those screens if needed. 

All in all, I think I still favor the push approach we originally mapped out. Curious if others agree.

**Comment by JmillsExpensify** [comment]
Another alternative I'd mention, is that we could easily remove one additional page and consideration by limiting the date picker to a year picker. After all, it's super quick to cycle 6 months one direction of the other. I also think that's fine, because this flow is going to be touched exactly once by users in the app. Your birthday never changes, and similarly, the date of incorporation of your company never changes. We can always add another page in the future for month too, I just don't think it's required to start, and there is no material impact to the user experience.

**Comment by shawnborton** [comment]
Oh interesting... I wonder if we could simplify things even further and use something like this? 
<img width="831" alt="image" src="https://user-images.githubusercontent.com/2319350/220385978-4b95b01e-16a4-4b78-b0bf-5c8bba053e96.png">


At the same time... for a birthdate, you'd be forcing the user to click as many times as they are old. So for me, I'd have to hit that back arrow 35 times until I got to the right year!

**Comment by mountiny** [comment]
@Beamanator I think the Onyx key approach is what we should/ would have to do, but I also think we would have to have Onyx key for each of the datepickers/ forms if we want to make sure the progress is saved. AFAIK for forms like bank account set up we are saving the progress and data inputted in Onyx so user can come back to it. The incorporation date would be lost in case we would go and change some other datepicker (as they would use same onyx key).

If we will have just a few Datepickers I think its fine to use such pattern.

@shawnborton I think removing the push to page only from `Month` would not solve the issue and as you said clicking through years is not great UX.

I think we have to make a decision now whether to follow the push to page pattern here as designed for consistent user experience, even with the technical complications and worse developer experience, which will most likely lead to more bugs with this component in future.

I would personally go for simplicity and promise of less issues in future and not use the push-to-page pattern here and keep this as one component. @JmillsExpensify @Beamanator @shawnborton what do you think? I think if we decide now, @ArekChr and @TMisiukiewicz could have PR ready by the EOW

**Comment by shawnborton** [comment]
I stated this above but I still feel like push-to-page is the best solution here. Curious what you think about my [comments here](https://github.com/Expensify/App/issues/14322#issuecomment-1438513571).

**Comment by JmillsExpensify** [comment]
I'm with Shawn. I also think we can simply the component and just push to page for year, keeping everything else within the component. That feels like a good compromise to simplify what we have and still remain consistent with other patterns throughout the app.

**Comment by mountiny** [comment]
> We already have an Onyx key for the form that the date value is being stored in, and we would have an inputID for that DatePicker so are you sure we'd need new onyx keys? In my mind we'd be able to reuse the onyx keys we already have 🤔

Ah, you will be right most likely, what was not clean to me from the summary you posted (but it was brief I know) is if the onyx data is specific to the form, hence the datepicker data saved is saved for each of the forms separately.

> What about having a full page "year selector" component that slides in just like a new page would, over the top of existing pages? This would eliminate the need for new routes, but I'm not so sure if the animation is easy to set up to look exactly like the navigation animation on all platforms?

I am not fan of this idea, I think that would just lead to more confusing code and bugs.

**Comment by TMisiukiewicz** [comment]
@Beamanator thanks for your input! We are currently implementing the first version of push-to-page pattern for a year picker and at first glance it looks like our concerns about depending heavily on the navigation events seems correct, anyway we managed to make it work pretty well and we are now thinking about making it as easy as possible to implement whenever DatePicker is used. 

> What about having 1 generic "year selector" page
We'd probably have to store settings (like which form the year selector would save to, which page to navigate back to (on back, on close, on save), min/max date validation, etc) in Onyx
Also users probably shouldn't be able to navigate directly to this page so we'd potentially want to route them elsewhere?

I think our goal should be achieving exactly something like this. Let's say if there is no param with a name of the screen the user was previously, it should automatically redirect somewhere else e.g. Homescreen. With this param set, we would always know where we should go back after selecting a date/going back, even if the user navigates to this page directly. 

If i understand correctly, the Onyx is used in the forms to store draft value of the input. IMO we shouldn't use it to keep the Year value (which is kind of a temporary) and instead pass the two params to the screen where to go back: year and input ID. Thanks to that, we will automatically know which DatePicker we should open after going back to the screen and update the Onyx state only after selecting the correct date by pressing the day. I think not selecting a full date should mean no change in Onyx.

Quick demo of what we were able to achieve with push-to-page for now:

https://user-images.githubusercontent.com/13985840/220907214-e14f446d-a28a-4554-8923-65706f92353f.mov

As you can see it does exactly what I mentioned earlier - updating the year only internally for a calendar and not updating a form value unless the full date is selected

**Comment by JmillsExpensify** [comment]
Just for the sake of clarity, if we go the route of only pushing to page for the year, wouldn't the component look more like this? 
<img width="264" alt="Screenshot 2023-02-23 at 23 12 59" src="https://user-images.githubusercontent.com/1127863/221097533-afb5e2e6-dda6-448a-9871-d10226769929.png">

**Comment by Beamanator** [comment]
@TMisiukiewicz that's looking pretty great so far! As mentioned by @JmillsExpensify it looks like there's a few small change y'all are still working on (a.k.a. moving the Year button to the left side, moving the Month to the right) but the POC looks great!

One possibly quick question: It looks like when you're on the Year page and you select a year and are navigated back to the page with the Date Picker, it looks like the calendar section of the page takes a second to load - is that because we have to calculate all of the days to show in the picker & which numbers they should have? Just wondering if we could have some kind of placeholder border or something that could give us less of a "nothing's here... OH WAIT the calendar is here" visual.

**Comment by shawnborton** [comment]
Oh nice, I like your [mockup here](https://github.com/Expensify/App/issues/14322#issuecomment-1442821453) @JmillsExpensify - and having the year on the left matches the date format we require, nice!

**Comment by TMisiukiewicz** [comment]
> One possibly quick question: It looks like when you're on the Year page and you select a year and are navigated back to the page with the Date Picker, it looks like the calendar section of the page takes a second to load - is that because we have to calculate all of the days to show in the picker & which numbers they should have? Just wondering if we could have some kind of placeholder border or something that could give us less of a "nothing's here... OH WAIT the calendar is here" visual.

Hi @Beamanator sorry for responding late but i was on a sick leave on Friday. The reason is we are using Popover component combined with the Calendar, and it's based on a Modal, which appears on top of the application. It means it gets visible on top of the app unexpectedly when the animation for navigation happens. So we have to make it visible right after transition ends. The previous video is from the debug mode - it looks smoother in the production build:

https://user-images.githubusercontent.com/13985840/221511178-20611658-b2ae-43dc-b418-558283fac7d0.mov

Anyway if it's still not looking the best let me know and I'll try to figue out if there is anything more I can do about this.

For now only some small improvements left and some code refactoring to make it easier to reuse in the future

**Comment by Beamanator** [comment]
@TMisiukiewicz Oh wow sorry for the late response, I read your comment yesterday, ate a sandwich, then forgot to come back to respond 🙃 🥪 

I think that production build looks great, I have no concerns about that anymore - thanks for sharing the explanation and what the production build looks like! 👍

**Comment by mountiny** [comment]
Great job @TMisiukiewicz 🎉

**Comment by TMisiukiewicz** [comment]
What do you think about having ✖️ icon on iOS and Android to close the calendar? On web it's pretty straightforward to click anywhere outside the calendar to close it but on mobile this kind of implementation gets a little bit tricky. Here's what i am thinking of:
![image](https://user-images.githubusercontent.com/13985840/222104698-7ee4c614-3160-4a90-87e7-c64f280c94a9.png)

**Comment by Beamanator** [comment]
> I wanted to check if not using existing popover would make more sense here because I still felt a bit insecure about all the events that would go under the hood and the need of tracking a lot of informations we could avoid. Created my own wrapper for calendar and basically achieved the same result with less code, less workarounds and easier to maintain code. E.g. the problem with transition I mentioned above does not exist anymore. Also managed to create just one screen for year picker so we would not end up with creating it each time DatePicker is used. Only some small improvements left - should go into our internal code review tomorrow and after that i will work on the PR checklist - this can be tracked in #15343 (it's a Draft for now)

@TMisiukiewicz that sounds fantastic, I would definitely say having less code that has less workarounds & easier to maintain is 💯 % an amazing reason to go that route! Nice work! 👍 

> What do you think about having ✖️ icon on iOS and Android to close the calendar? On web it's pretty straightforward to click anywhere outside the calendar to close it but on mobile this kind of implementation gets a little bit tricky.

So just to be clear, I have two things in my head: First, I think we should strive to make this calendar / date picker look the same on all platforms - so if we add an ✖️ icon in native, it should also be visible in web / desktop. That being said, on this specific page (date of birth) I'm not sure I see the need to close the picker, since there's nothing else to view 🤔  If the date picker is showing over other form inputs (on a bigger form) then I definitely like the idea of being able to close the picker with the ✖️ icon. So maybe just add a prop to show / hide the close button? 

Also interested in @shawnborton 's opinions on the proposed ✖️  icon to close the picker! 👍

**Comment by mountiny** [comment]
I am not that persuaded about the X button as its going to be harder to achieve this cross-platform. 

@Beamanator where would the date picker be on a page with other inputs? Woudnt we always push to page where there would be only the date picker? Hence covering other inputs would not be such a big issue?

I would vote for skipping this for now and we can add this functionality later if we see need for it.

**Comment by shawnborton** [comment]
I agree with the comments thus far - let's pass on implementing a "x" for now.

**Comment by Beamanator** [comment]
I'm down to skip the ✖️ for now 👍  though I don't know how it's hard to make it cross platform :D

**Comment by TMisiukiewicz** [comment]
two new copies need to be reviewed:
```
yearPickerPage: {
        year: 'Year',
        selectYear: 'Please select a year',
    },
```

As we want to use Year picker screen for other date pickers as well in the future, copies here need to be generic

![image](https://user-images.githubusercontent.com/13985840/222388988-bd2d2555-2ffc-4ade-a7af-e29cae051ce2.png)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jamesdeanexpensify (`Waiting for copy`), see https://stackoverflow.com/c/expensify/questions/7025/ for more details.

**Comment by Beamanator** [comment]
@TMisiukiewicz What about putting the "Please select a year" copy (pending approval) as the Search input label?

**Comment by s77rt** [comment]
@mountiny Can you assign me here as well?

**Comment by jamesdeanexpensify** [comment]
OK, so you just want copy for three screens, is that accurate? [This screen](https://github.com/Expensify/App/issues/14322#issuecomment-1451567093), and then the same for month and day?

If that's the case, I would actually keep the header the same "Date of birth" for all three screens when you click into them, and then for the subheader for each:
- Please select a year.
- Please select a month. 
- Please select a day.

**Comment by TMisiukiewicz** [comment]
@jamesdeanexpensify Actually it's just for the year picker screen mentioned above. The thing is if the Datepicker component will be used in any other form, we'd have to define what header and subheader to display for each one of them. To avoid that I thought about the generic messages here. WDYT?

**Comment by jamesdeanexpensify** [comment]
Ah, I get it. Thanks for the explanation! In that case, I think "Please select a year." works for now and we can always update if anyone gets confused or brings up an issue.

**Comment by MelvinBot** [comment]
@shawnborton, @mananjadhav, @Beamanator, @TMisiukiewicz, @s77rt, @jamesdeanexpensify, @ArekChr Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by s77rt** [comment]
Still working on PR. Almost done

**Comment by MelvinBot** [comment]
⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/16375)

If you are the assigned CME please investigate whether the linked PR caused a regression and leave a comment with the results.

If a regression has occurred and you are the assigned CM follow the instructions [here](https://stackoverflow.com/c/expensify/questions/13647).

If this regression could have been avoided please consider also proposing a recommendation to the PR checklist so that we can avoid it in the future.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.88-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15343


If no regressions arise, payment will be issued on 2023-03-31. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by mountiny** [comment]
@jamesdeanexpensify I think it should be @mananjadhav and @s77rt who should be both paid $1000 for testing and review of this one

**Comment by jamesdeanexpensify** [comment]
Ah, I think the BugZero team pays out for these, and I'm not on that team! We might just have to add the BugZero label?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`NewFeature`), see https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=BugZero%20process%20steps%20for%20feature%20requests for more details.

**Comment by mountiny** [comment]
Oops, sorry James.

That will be then this information https://github.com/Expensify/App/issues/14322#issuecomment-1483209316 for @Christinadobrzyn on 31st March when payments are due.

**Comment by mananjadhav** [comment]
@mountiny @Beamanator I think @s77rt deserves a raise here. They not only reviewed the PR but also provided technical inputs in solving some challenges. Overall the PR had 29 files and it went through 200+ conversations to resolve all the issues, etc.

**Comment by Beamanator** [comment]
That aaaaaabsolutely sounds super fair to me 👍

**Comment by s77rt** [comment]
Thanks @mananjadhav and @Beamanator But I'm actually happy with the flat $1K as I have also reviewed some pretty quick and easy PRs for the same compensation. So things even out eventually.

**Comment by mountiny** [comment]
Up to you @s77rt 🙌

**Comment by Christinadobrzyn** [comment]
Just checking on this! It sounds like we're good to pay? I don't see that a bonus is involved with this one but it sounds like we're thinking of maybe a little extra for @s77rt?

How about: 
@mananjadhav - $1000
@s77rt - $1500 (treat it like a PR bonus?)

or should we @mananjadhav the same? not exactly sure how much work was done on this. Just let me know!

**Comment by Christinadobrzyn** [comment]
Just catching up here because I'm new to the New Feature label. It looks like this job is complete so I think all we need to do is pay in Upwork? Is there something else that needs to happen as this relates to the New Feature label?

**Comment by mountiny** [comment]
Seems like both C+ were paid so we are good to close this, @mananjadhav @s77rt can you confirm?

**Comment by mananjadhav** [comment]
@mountiny Not paid yet.

**Comment by s77rt** [comment]
Not yet.
@Christinadobrzyn It's $1000 for me too.

**Comment by mountiny** [comment]
@Christinadobrzyn so maybe we will have to create new jobs for this.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01af33f93559a520a9

**Comment by MelvinBot** [comment]
Current assignees @mananjadhav and @s77rt are eligible for the Internal assigner, not assigning anyone new.

**Comment by Christinadobrzyn** [comment]
Hired @s77rt and @mananjadhav in Upwork - https://www.upwork.com/ab/applicants/1642945974459355136/job-details

Paid @mananjadhav $1000
@s77rt can you accept my offer in [Upwork](https://www.upwork.com/jobs/~01af33f93559a520a9) and I'll pay you!

**Comment by s77rt** [comment]
@Christinadobrzyn Accepted

**Comment by Christinadobrzyn** [comment]
Paid @s77rt $1000 in upwork

Closed the job in Upwork. 

@mananjadhav do you see any reason to create a regression test for this? If you think it's a good idea, can you send me your suggested regression test steps and I'll send it over to our QA team!

**Comment by mountiny** [comment]
@Beamanator Good point, I think we should add some regression steps for this component. They should be quite thorought.

Both of you @mananjadhav @s77rt would you be able to compile some of the steps from the PR. There is lots to test, but probably the main flows, choosing year, month, going there and back.

**Comment by s77rt** [comment]
@mountiny Do we add regression tests (manual ones) even if we have automated tests? The main flow is already covered by automatic tests here.

**Comment by mountiny** [comment]
I  mean we already had some regressions from it, we dont have to add regression steps for what the automated test covers, but for other things yes.

Or add those tests as automated ones.

**Comment by s77rt** [comment]
**Regression Test Proposal**
1. Open App
2. Go to Settings > Profile > Personal details > Date of birth
3. Hover the calendar icon
4. Verify the mouse cursor is a pointer
5. Note the current date shown in the input
6. Click the calendar icon
6. Verify the calendar shows up in the noted date
7. Click on current year
8. Verify you can't type more than 4 digits
9. Select any year (different than current one)
10. Close calendar (click outside)
11. Reopen calendar
12. Verify the calendar still shows up in the same previous noted date
13. Hover any day
14. Verify the day is styled with a light background
15. Choose any day
16. Go back (without saving changes)
17. Go to DOB page again
18. Open calendar
19. Verify the calendar still shows up in the same previous noted date
20. Select a date
21. Save changes
22. Verify DOB now shows the new selected date

**Comment by mountiny** [comment]
@s77rt we have updated the behaviour in [this PR](https://github.com/Expensify/App/pull/16779)  so that the calendar is open when navigating to this page, so we should update the steps to reflect that

**Comment by s77rt** [comment]
@mountiny Thanks. While trying to update the regression proposal I found that the linked PR introduced a regression https://expensify.slack.com/archives/C049HHMV9SM/p1680625096406979. It should be a quick fix though https://github.com/Expensify/App/pull/16779/files#r1157491589

**Comment by Christinadobrzyn** [comment]
Just to confirm, we're waiting for a revised Regression Test, correct? @s77rt

**Comment by s77rt** [comment]
@Christinadobrzyn Yes, but I'm waiting for https://github.com/Expensify/App/pull/16941 to be merged first.

**Comment by shawnborton** [comment]
Not overdue

**Comment by mountiny** [comment]
> Is the long-term plan to eliminate our use of the "old" DatePicker that uses native controls from @react-native-community/datetimepicker under the hood on native and is just a TextInput on web?

Yes, the long term plan is to eliminate this everywhere, there is no tracking issue for this because afaik its basically in connection to the forms/ push-to-page refactor. As we refactor those forms to push-to-page, we would use this new component for any date pickers.

> You can't type in the text input, you have to use the calendar to pick a date. Is this expected? If so, why?

This is intended and I believe its mainly to unify the experience. Also to [avoid different formats](https://github.com/Expensify/App/issues/14322#issuecomment-1431857358). @Beamanator might know better in terms of some predesign for this (I kinda stumbled upon this and stuck around helping moving this forwards)

> There's no push-to-page input for the month picker – you just have to cycle through using the < and >. This seems fine, but is just kind of inconsistent with how it works for year. Saw the mockups above included push-to-page for months and wanted to make sure that we landed where we wanted.

That is intended as well, it was a design choice made I think in [this comment](https://github.com/Expensify/App/issues/14322#issuecomment-1438651374), basically we are simplifying it and also we need to consider this is very rare component. You sat your DOB just once and then in other places the date component is used is rare flow

**Comment by roryabraham** [comment]
> Manual input in the top-line date field shouldn't be allowed. Otherwise we have to handle all kinds of date formats

Not necessarily true – we can still enforce the date format

**Comment by s77rt** [comment]
**Regression Test Proposal**

1. Open App
2. Go to Settings > Profile > Personal details > Date of birth
3. Note the written date in the input
4. Verify the calendar is shown in that noted date
5. Click on current year
6. Verify you can't type more than 4 digits
7. Select any year
8. Verify the calendar is still shown
9. Click < or > to change the month
10. Verify the month is changed accordingly
11. Hover any day
12. Verify the day is styled with a light background
13. Choose any day
14. Go back (without saving changes)
15. Go to DOB page again
16. Verify the calendar still shows up in the same previous noted date
17. Select a date
18. Save changes
19. Verify DOB now shows the new selected date

**Comment by mountiny** [comment]
I think this looks good for now and we should not hold this issue for any future updates. We should update this regression test rather if we change anything about the behaviour.

cc @Christinadobrzyn would you be able to get the issue going to insert this into TestRail?

**Comment by Christinadobrzyn** [comment]
Yep! created Test Rail [GH](https://github.com/Expensify/Expensify/issues/275332) for QA. I think this GH can be closed?

**Comment by Christinadobrzyn** [comment]
I'll close this GH as I think it's complete but feel free to reopen if I've missed something.
